tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -11882,6 +11909,33 @@ bool hlsl::ShouldSkipNRVO(clang::Sema &sema, clang::QualType returnType,
   return false;
 }
 
+bool hlsl::HasLongVecs(const QualType &qt) {
+  if (qt.isNull()) {
+    return false;
+  }
+
+  if (IsHLSLVecType(qt)) {
+    if (GetHLSLVecSize(qt) > 4)
+      return true;
+  } else if (qt->isArrayType()) {
+    const ArrayType *arrayType = qt->getAsArrayTypeUnsafe();
+    return HasLongVecs(arrayType->getElementType());
+  } else if (qt->isStructureOrClassType()) {
+    const RecordType *recordType = qt->getAs<RecordType>();
```

Comment:
We could potentially avoid doing the record decl traversal at all if we take a bit from `DefinitionData` and cache it as we add members and base classes to the decl.

Reply:
It's an interesting idea. I didn't like the recursion and throwing in base classes just adds to the overhead.

Reply:
This is requiring a lot more additional code than I am inclined to put in here. In particular, various template parameters are not instantiated until after the checks are done now, so the DefintionData is set way too late. We have some forcing of instantiation, but only if the parameter is a template itself. @llvm-beanz Will you still consider approving the change without this?

Reply:
I'm a bit surprised by this, DefinitionData tends to get updated as members get added into records, so I would have expected it to be sufficiently easy to update this in the `addMember` function.

I don't consider this a blocker if it is more complicated than I expected.

Reply:
It's mainly the ConstantBufer<LonVec> usage where the template parameter is already weird and doesn't get addressed in the usual order of things. Other usages might be easier, but that was the one I started with.

Reply:
Done